File: Objective-C/B/BinSearch_Iterative.m
#import <Foundation/Foundation.h>

@interface NSArray (BinarySearch)
- (NSInteger) binSearch:(id)key;
@end

@implementation NSArray (BinarySearch)
- (NSInteger)binSearch:(id)key {
    NSInteger lowerIndex = 0;
    NSInteger upperIndex = [self count] - 1;
    while (lowerIndex <= upperIndex) {
        NSInteger midIndex = lowerIndex + (upperIndex - lowerIndex) / 2;
        id centVal = self[midIndex];
        
        switch([centVal compare:key]) {
            case NSOrderedAscending:
                lowerIndex = midIndex + 1;
                break;
            case NSOrderedDescending:
                upperIndex = midIndex - 1;
                break;
            case NSOrderedSame:
                return midIndex;
        }
    }
    return NSNotFound;
}
@end

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        
        NSArray *values = @[@1, @2, @3, @4];
        NSLog(@"Value is in array element: %li", [values binSearch:@4]);
    }
    return 0;
}

File: Swift/B/BinSearch_Iterative.swift
import Foundation

func binarySearch<T: Comparable> (xs: [T], x: T) -> Int? {
  var (lowerIndex, upperIndex) = (0, xs.count - 1)
  
  while lowerIndex <= upperIndex {
    switch (lowerIndex + upperIndex) / 2 {
      case let midIndex where xs[midIndex] > x: upperIndex = midIndex - 1
      case let midIndex where xs[midIndex] < x: lowerIndex = midIndex + 1
      case let midIndex: return midIndex
    }
  }
  return nil
}

let arr = [1, 2, 3, 4, 5, 6, 7, 9, 11]

if let binSearch = binarySearch(xs: arr, x: 1) {
	print("Element is present in array: ", binSearch)
} else {
	print("Element is not present in the array!")
}
